### DISCUSSION TITLE
use simpIncScr in other extension

### DISCUSSION THREAD

**Discussion by pat123456** [spakin/SimpInkScr#141] [Q&A] [answered]
Hello,
I would like to use SimpIncScr functions in an other extension code, is it possible ?
Thanks for help

**Comment by spakin** [comment]
I believe it _should_ be possible, but I admit I haven't given this much thought.  Try including a
```Python
from simpinkscr import *
```
in the other extension code and seeing if that provides access to the Simple Inkscape Scripting functions you'd like to use.  Does that work?

**Comment by pat123456** [comment]
Hello, thank you for your answer, I did a try but get this error :

xxxx/extensions/hello/Hello.py",` line 25, in effect
    circle((100, 100), 200)
  xxxx/extensions/org.inkscape.extension.29361/simple_inkscape_scripting/simpinkscr/simple_inkscape_scripting.py", line 2725, in circle
    return SimpleObject(obj, transform, conn_avoid, clip_path, mask,
           ^^^^
  xxxx/extensions/org.inkscape.extension.29361/simple_inkscape_scripting/simpinkscr/simple_inkscape_scripting.py", line 427, in __init__
    _simple_top.append_obj(self)
    ^^^^
AttributeError: 'NoneType' object has no attribute 'append_obj'

**Reply by spakin** [reply]
Thanks for reporting back.  Could you please post your `Hello.py` so I can try to reproduce the problem?  Hopefully, there's a simple fix.

**Reply by pat123456** [reply]
here is the code (I'm novice in python, may be there is a better/simpler way get the path of simpinkscr script ?)

```
import os
import sys
import inkex

parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sis_path = os.path.join(parent_dir, 'org.inkscape.extension.29361/simple_inkscape_scripting/simpinkscr')
sys.path.append(sis_path)

from simple_inkscape_scripting import *

class Hello(inkex.GenerateExtension):

	def effect(self):
		circle((100, 100), 200)

if __name__ == "__main__":
	Hello().run()
```

**Reply by spakin** [reply, marked as answer]
Thanks for the code sample.  I was able to use it to reproduce the problem.

What's happening is that Simple Inkscape Scripting initializes its data structures during execution of its `effect` method.  But Simple Inkscape Scripting's `effect` method is never called; only Hello's `effect` method is called.  Unfortunately, the solution is not as simple as explicitly invoking Simple Inkscape Scripting's `effect` method because `effect` relies on variables that inkex initializes when invoking the extension.  But inkex isn't invoking the Simple Inkscape Scripting extension; it's invoking the Hello extension.  Ugh.

As a horribly ugly solution, you can manually initialize just enough of Simple Inkscape Scripting's data for an explicit call to `effect` to succeed.  Here's what that might look like:
```Python
import argparse
import os
import sys
import inkex

sys.path.append(os.path.abspath('simple_inkscape_scripting'))

from simpinkscr import *
from simpinkscr.simple_inkscape_scripting import SimpleInkscapeScripting

class Hello(inkex.GenerateExtension):

        def effect(self):
                # Manually initialize Simple Inkscape Scripting.
                sis = SimpleInkscapeScripting()
                sis.svg = self.svg
                sis.options = argparse.Namespace()
                sis.options.user_args = []
                sis.options.encoding = ''
                sis.options.py_source = None
                sis.options.program = None
                sis.effect()
                
                circle((100, 100), 200)

if __name__ == "__main__":
        Hello().run()
```

If you place your INX file at the same level of your directory hierarchy as the `simple_inkscape_scripting` directory, a straightforward `sys.path.append(os.path.abspath('simple_inkscape_scripting'))` should work, as used in the above.

If your extension consists entirely of code that can run from Simple Inkscape Scripting, you may be interested in a page I recently added to the documentation: [Derived extensions](/spakin/SimpInkScr/wiki/Derived-extensions).  The approach presented there is to put all your Simple Inkscape Scripting code in an INX file and not create your own Python extension class at all.  This is a lot cleaner than manually initializing Simple Inkscape Scripting's global state, but it doesn't give you the same degree of control that a do-it-yourself extension does.

**Comment by pat123456** [comment]
Thank you very much for your response and research! It works very well.

I hadn't noticed that you added the "derived extensions" page following the question I had deleted. I removed the question because it seemed like it was covering two somewhat different topics in the same discussion. Thank you for responding to it; the solution is also interesting.

Perhaps it would be helpful, for the direct use of simpIncScr in extension code, to have a special version with only the useful functions (circle, rect, bounding_box, all_shapes...) and without the unnecessary ones (svg_root, extension...) and problems of initialisation. Essentially a simple library rather than an extension. But this is just a idea, not a request, as it would certainly involve a lot of work.

Thank you again for your attention, and have a good day!

**Reply by spakin** [reply]
> I hadn't noticed that you added the "derived extensions" page following the question I had deleted. I removed the question because it seemed like it was covering two somewhat different topics in the same discussion. Thank you for responding to it; the solution is also interesting.

I found your deleted question raised an inspiring idea.  I hope people find the new wiki page useful.

> Perhaps it would be helpful, for the direct use of simpIncScr in extension code, to have a special version with only the useful functions (circle, rect, bounding_box, all_shapes...) and without the unnecessary ones (svg_root, extension...) and problems of initialisation. Essentially a simple library rather than an extension. But this is just a idea, not a request, as it would certainly involve a lot of work.

It wouldn't take much work simply to drop the "Manually initialize Simple Inkscape Scripting" stanza from my previous answer into a function that Simple Inkscape Scripting exports, say `init_simpinkscr_effect`.  I don't believe that would expose names like `svg_root` because they're not listed in [`__init__.py`](https://github.com/spakin/SimpInkScr/blob/master/simpinkscr/__init__.py).

This would not be as transparent as a library as you'd still have to invoke `init_simpinkscr_effect` from your `effect` method.  But your extension won't have to be aware of Simple Inkscape Scripting's internals so that could be considered at least a step forward.

**Reply by svalade** [reply]
First of all, many thanks for this awesome repo, it's has been really helpful!

My question is similar to this discussion, and to the answers provided. It feels trivial, and yet I cannot get a working solution.

I have created a python script which I call from the Simple Inkscape Scripting dialog box. In this script I have a number of functions, which use Simple Inkscape Scripting functions. I would like to be able to reuse my functions in other scripts, therefore I'd essentially like to create a module in which I can import simpinkscr functions (circle, connector, style, group, etc.).

I've tried to manually initialize Simple Inkscape Scripting as you showed in your answer, in order to avoid the error " AttributeError: 'NoneType' object has no attribute 'append_obj'", but without success.

Any guidance would be greatly appreciated!

**Reply by spakin** [reply]
> I have created a python script which I call from the Simple Inkscape Scripting dialog box. In this script I have a number of functions, which use Simple Inkscape Scripting functions. I would like to be able to reuse my functions in other scripts, therefore I'd essentially like to create a module in which I can import simpinkscr functions (circle, connector, style, group, etc.).

Try `read`ing and `exec`ing your module instead of `import`ing it.  (`import` behaves in unexpected ways when invoked from `exec`, which is what Simple Inkscape Scripting uses to run a script.)

As a concrete example, put the following code in a module called `helpers.py`:
```Python
def donut(c, r1, r2, **kwargs):
    'Draw a donut.'
    circ1 = circle(c, r1, **kwargs).to_path()
    circ2 = circle(c, r2, **kwargs).to_path().reverse()
    circ1.append(circ2)
```
Then, from each script that wants to reuse `donut`, "import" `helpers.py` as follows:
```Python
helpers_py = os.path.join(inkex.utils.get_user_directory(), "helpers.py")
with open(helpers_py) as helpers:
   exec(helpers.read())
```
I stored `helpers.py` in my extensions directory (`~/.config/inkscape/extensions/` on my Linux system) so the above finds it using `inkex.utils.get_user_directory()`, but you can hard-wire the path if you prefer.

At this point, the `donut` function is available to the script as if the contents of `helpers.py` had been copied and pasted into the script:
```Python
donut((100, 100), 50, 40, fill='green')
```

**Reply by svalade** [reply]
This works brilliantly! Many many thanks for taking the time to reply, and for providing this very useful solution. Awesome

**Comment by spakin** [comment]
As of d4aa42fbf42013af7dcd83d18b226a6c2279324a, Simple Inkscape Scripting provides more explicit support for exposing its functions to other extensions.  See [Leveraging Simple Inkscape Scripting in extensions](https://github.com/spakin/SimpInkScr/wiki/Derived-extensions#leveraging-simple-inkscape-scripting-in-extensions) in the wiki for documentation.
